siteorigin-panels.php

```diff
@@ -343,20 +335,83 @@ public function generate_post_content( $content ) {
 		return $content;
 	}
 	
+	/**
+	 * Generate an excerpt for the current post, if possible.
+	 *
+	 * @param $text
+	 *
+	 * @return mixed|string
+	 */
+	public function generate_post_excerpt( $text ) {
+		global $post;
+		if ( empty( $post ) && ! in_the_loop() ) {
+			return $text;
+		}
+		
+		$post_id = $this->get_post_id();
+		
+		// Check if this post has panels_data
+		$panels_data = get_post_meta( $post_id, 'panels_data', true );
+		if ( $panels_data && ! empty( $panels_data['widgets'] ) ) {
+			
+			foreach ( $panels_data['widgets'] as $widget ) {
+				$panels_info = $widget['panels_info'];
+				if ( $panels_info['grid'] > 1 ) {
+					// Limiting search for a text type widget to the first two PB rows to avoid having excerpt content
+					// that's very far down in a post.
+					break;
+				}
+				if ( $panels_info['class'] == 'SiteOrigin_Widget_Editor_Widget' || $panels_info['class'] == 'WP_Widget_Text' ) {
+					$text = $widget['text'];
+					// This is all effectively default behavior for excerpts, copied from the `wp_trim_excerpt` function.
+					// We're just applying it to the first text type widget's content.
+					$raw_excerpt = $text;
```

Comment:
Sometime I just thought of here. What if the user has spread their opening content across a few text widgets. This block should just keep adding to the `$raw_excerpt` variable until its length is >= `$excerpt_length`. Once that happens, it should break out of the `foreach` loop.

All the rest of the processing should be applied to `$raw_excerpt` after the `foreach` loop.

Reply:
Hmmm, yeah good point. Users could use various different configurations of text widgets, so perhaps we need to allow them to specify exactly which widgets they might want as excerpts with a checkbox in Widget Settings?

Alternatively, perhaps we should give them a way to switch the auto-excerpt functionality? Just slightly concerned that we might end up using things as excerpts that don't make sense. E.g. what if a user is using the Editor Widget just for headings, or images, or some other HTML?

Reply:
I think just relying on any text-like widgets is good enough. If anything is really weird with the excerpt, then the solution is to manually write one.

The auto-excerpt should be 100% automatic, so I'd be against adding any settings/checkboxes. What we're trying to do here is make the automaticness slightly more intelligent.

Reply:
Ok, cool. Then I'd just need to change the behavior slightly. Need to run our filter earlier than the default filter and then if there's an existing manual excerpt, it will just return that.